server/ocsp.go

```diff
@@ -0,0 +1,377 @@
+func (srv *Server) NewOCSPMonitor(tc *tls.Config) (*tls.Config, *OCSPMonitor, error) {
+	opts := srv.getOpts()
+	oc := opts.OCSPConfig
+	tcOpts := opts.TLSConfigOpts
+	if oc.Mode == OCSPModeNever {
+		return tc, nil, nil
+	}
+	cert, err := tls.LoadX509KeyPair(tcOpts.CertFile, tcOpts.KeyFile)
+	if err != nil {
+		return nil, nil, err
+	}
+	cert.Leaf, err = x509.ParseCertificate(cert.Certificate[0])
+	if err != nil {
+		return nil, nil, err
+	}
+	if oc.Mode == OCSPModeAuto && !hasOCSPStatusRequest(cert.Leaf) {
+		// "status_request" flag not set. No need to do anything.
+		return tc, nil, nil
+	}
+
+	// TODO: Add OCSP 'responder_cert' option in case CA cert not available.
+	caCert := tcOpts.CaFile
+	issuer, err := getOCSPIssuer(caCert, cert.Certificate)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Callbacks below will be in charge of returning the certificate instead,
+	// so this has to be nil.
+	tc.Certificates = nil
+
+	mon := &OCSPMonitor{
+		srv: srv,
+		hc:  &http.Client{Timeout: 30 * time.Second},
+		// FIXME: Make reloadable aware, otherwise a change in certificates
+		//  could make the staples stale.
+		Leaf:   cert.Leaf,
+		Issuer: issuer,
+	}
+
+	// Get the certificate status from the memory, disk, then remote OCSP responder.
+	_, resp, err := mon.getStatus()
+	if err != nil {
+		return nil, nil, fmt.Errorf("bad OCSP status update: %s", err)
+	}
+	if err == nil && resp.Status != ocsp.Good && (oc.Mode == OCSPModeAlways || oc.Mode == OCSPModeMust) {
+		return nil, nil, fmt.Errorf("found existing OCSP certificate status: %s", ocspStatusString(resp.Status))
+	}
+
+	// GetCertificate returns a certificate that's presented to a
+	// client.
+	tc.GetCertificate = func(info *tls.ClientHelloInfo) (*tls.Certificate, error) {
+		raw, _, err := mon.getStatus()
+		if err != nil {
+			return nil, err
+		}
+
+		return &tls.Certificate{
+			OCSPStaple:                   raw,
+			Certificate:                  cert.Certificate,
+			PrivateKey:                   cert.PrivateKey,
+			SupportedSignatureAlgorithms: cert.SupportedSignatureAlgorithms,
+			SignedCertificateTimestamps:  cert.SignedCertificateTimestamps,
+			Leaf:                         cert.Leaf,
+		}, nil
+	}
+
+	// GetClientCertificate returns a certificate that's presented to a
+	// server.
+	tc.GetClientCertificate = func(info *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+		return &cert, nil
+	}
+	return tc, mon, nil
+}
+
+func hasOCSPStatusRequest(cert *x509.Certificate) bool {
+	tlsFeatures := asn1.ObjectIdentifier{1, 3, 6, 1, 5, 5, 7, 1, 24}
+	const statusRequestExt = byte(5)
+	const tlsFeaturesLen = 5
+
+	// Example Value: [48 3 2 1 5]
+	// Documentation:
+	// https://tools.ietf.org/html/rfc6066
+
+	for _, ext := range cert.Extensions {
+		if !ext.Id.Equal(tlsFeatures) {
+			continue
+		}
+		if len(ext.Value) != tlsFeaturesLen {
+			continue
+		}
+		return ext.Value[tlsFeaturesLen-1] == statusRequestExt
```

Comment:
This needs a HUGE FIXME comment on it.
The value of the extension identified by `tlsFeatures` is an ASN.1 sequence of integers.
Here you've checked that the payload has a length matching the byte length if there's just one small integer in the list and checked that the last field is that small integer.

A `tlsFeatures` extension with multiple features in it would cause a parse failure here.

https://play.golang.org/p/kCgFUCg2gYE is an example of how to do this right; specifically:
```
var intSeq []int
rest, err := asn1.Unmarshal(rawBytes, &intSeq)
```

Once you have the `intSeq` then you can iterate it looking for the `statusRequestExt` value.

Reply:
Awesome, here's the updated version: https://github.com/nats-io/nats-server/pull/2210/commits/30e741e8af3ea4297a4bcc7cb2c9655bf5cfa5e5